### ISSUE TITLE
setuptools.file_finders entrypoint is a giant time sink

### ISSUE THREAD

**Issue by tlandschoff-scale** [pypa/setuptools#1082]
Yesterday I fought with getting my build running on a Gitlab CI project. The build failed unhelpfully with this message:

```
$ strace -o toxfail.trace -f python2.7 /builds/scale-open-source/pyinstaller/setup.py sdist --formats=zip --dist-dir /builds/scale-open-source/pyinstaller/.tox/dist
running sdist
running egg_info
creating PyInstaller.egg-info
writing requirements to PyInstaller.egg-info/requires.txt
writing PyInstaller.egg-info/PKG-INFO
writing top-level names to PyInstaller.egg-info/top_level.txt
writing dependency_links to PyInstaller.egg-info/dependency_links.txt
writing entry points to PyInstaller.egg-info/entry_points.txt
writing manifest file 'PyInstaller.egg-info/SOURCES.txt'
error: [Errno 2] No such file or directory
```

As you can see from the command I was already resorting to using strace to find out what's going wrong here. In the end it turned out that some freaking file_finders implementation was installed and failed to find the git executable (which is not installed in the container used for the build).

I run into variations of this again and again, most of the time only when checking a package I just created via `python setup.py sdist` via `tardiff` against a previous version. Sometimes the size of the package mysteriously doubles - then I can be sure that some dependency pulled in either `setuptools_scm` or `hgdistver` or some other package messing with the file lists.

In my opinion this `file_finders` feature is a bug and must die. Installing a package from PyPI should not cause additional files to be packaged. I'd be fine with a mechanism to explicitly request this from setup.py. Using entry points for this make this just a hidden uber global variable for me.

### Workaround

In desperation I added the following command to my CI builds:

```python
python -c 'import pkg_resources;print("\n".join(d.key
        for d in pkg_resources.working_set
        if "setuptools.file_finders" in d.get_entry_map()))' \
    | while read p; do pip uninstall --yes $p; done
```

And magically all problems seem to disappear.

**Comment by jaraco** [comment]
Ignore that aforementioned commit. I think it was intended for another issue.

You're right - the file finders functionality can be surprising, especially because the behavior is injected by the presence of third-party libs.

I do recommend that these file finder libraries only be installed transiently, as part of a build of a single package. Once PEP 518 lands, it should better guarantee that dependencies like these aren't brought into the build of your package by the build of other packages in the same session.

But as far as deprecating and removing this functionality, I don't think that can happen. It provides too much value and has too many users relying on it. The advantages of the feature outweigh the harm of issues like these.

I don't find that in my environment that these file finders libraries get installed permanently. I suggest you track down what is causing these packages to be installed and see if you can convince them not to install those packages.

Hope that helps.
